### PR TITLE
Feature/cache detail

### LINKED ISSUE #56: [FR]: Cache movie details in local database
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Related app

All - Android

### Describe the problem

Cache movie and series details in local database to decrease API requests and speed up accessing it.

### Describe the solution

_No response_

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by hadi-norouzi**
### Description
Cache permanent movie details in the Room database.

### Checklist
- [x] Make sure to open a GitHub issue as a bug/feature request before writing your code! That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the linter (code style) passes
- [x] Appropriate docs were updated (if necessary)

Fixes #56

### CODE REVIEW

File: data/database/src/main/java/io/filmtime/data/database/DatabaseModule.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import dagger.Module
+import dagger.Provides
+import dagger.hilt.InstallIn
+import dagger.hilt.android.qualifiers.ApplicationContext
+import dagger.hilt.components.SingletonComponent
+import javax.inject.Singleton
+
+@Module
+@InstallIn(SingletonComponent::class)
+object DatabaseModule {
```
Comment (moallemi) [inline data/database/src/main/java/io/filmtime/data/database/DatabaseModule.kt:14]: Please make DatabaseModule `internal`

File: data/database/src/main/java/io/filmtime/data/database/FilmTimeDatabase.kt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package io.filmtime.data.database
+
+import androidx.room.Database
+import androidx.room.RoomDatabase
+
+@Database(entities = [MovieDetailEntity::class], version = 1)
+abstract class FilmTimeDatabase : RoomDatabase() {
```
Comment (moallemi) [inline data/database/src/main/java/io/filmtime/data/database/FilmTimeDatabase.kt:7]: Please make FilmTimeDatabase `internal`

File: data/database/src/main/java/io/filmtime/data/database/MovieDetailEntity.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package io.filmtime.data.database
+
+import androidx.room.Entity
+import androidx.room.PrimaryKey
+
+@Entity
+data class MovieDetailEntity(
```
Comment (moallemi) [inline data/database/src/main/java/io/filmtime/data/database/MovieDetailEntity.kt:7]: I'd suggest to use table name for db entities, so Renaming the kotlin model does not affect the sqlite DB queries and table names
```suggestion
@Entity(tableName = "movie_detail")
data class MovieDetailEntity(
```

File: data/tmdb-movies/src/main/java/io/fimltime/data/tmdb/movies/MovieDetailEntityExt.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import io.filmtime.data.model.VideoId
+
+fun MovieDetailEntity.toMovie(): VideoDetail {
+  return VideoDetail(
+    ids = VideoId(tmdbId = tmdbId, traktId = null),
+    title = name,
+    posterUrl = posterUrl,
+    coverUrl = coverUrl,
+    year = year,
+    genres = listOf(),
```
Comment (moallemi) [inline data/tmdb-movies/src/main/java/io/fimltime/data/tmdb/movies/MovieDetailEntityExt.kt:14]: if we want to really work offline we have to handle `genres` and `spokenLanguages` as well. If you do not to do it now, please add a TODO with an issue number

File: data/tmdb-movies/src/main/java/io/fimltime/data/tmdb/movies/TmdbMovieRepositoryImpl.kt [hunk 5]
```diff
@@ -6,29 +6,35 @@ import androidx.paging.PagingData
...
 import io.filmtime.data.model.GeneralError
 import io.filmtime.data.model.Result
 import io.filmtime.data.model.VideoDetail
 import io.filmtime.data.model.VideoThumbnail
 import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
 import javax.inject.Inject
 
-internal class TmdbMovieRepositoryImpl @Inject constructor(
+class TmdbMovieRepositoryImpl @Inject constructor(
```
Comment (moallemi) [inline data/tmdb-movies/src/main/java/io/fimltime/data/tmdb/movies/TmdbMovieRepositoryImpl.kt:19]: Why not `internal`? 🤔

File: data/tmdb-movies/src/main/java/io/fimltime/data/tmdb/movies/TmdbMovieRepositoryImpl.kt [hunk 6]
```diff
@@ -39,21 +45,25 @@ internal class TmdbMovieRepositoryImpl @Inject constructor(
...
+                  )
+                }
               }
             }
           }
         }
       }
     }
+
+    emit(result)
```
Comment (moallemi) [inline data/tmdb-movies/src/main/java/io/fimltime/data/tmdb/movies/TmdbMovieRepositoryImpl.kt:66]: We are not implementing offline first correctly here since you are always emitting the result of network call. We should only emit the network call result if local movie is null, also no need to even emit the success result if we observe the `getMovieByTmdbId`

File: gradle/libs.versions.toml [hunk 7]
```diff
@@ -70,6 +76,7 @@ com-android-library = { id = "com.android.library", version.ref = "agp" }
 kotlinx-serialization = { id = "org.jetbrains.kotlin.plugin.serialization", version.ref = "kotlin" }
 hilt-android = { id = "com.google.dagger.hilt.android", version.ref = "hilt-android" }
 spotless = "com.diffplug.spotless:6.20.0"
+ksp = { id = "com.google.devtools.ksp", version.ref = "ksp" }
```
Comment (moallemi) [inline gradle/libs.versions.toml:79]: You added the ksp dependency but you forgot to use it

File: data/tmdb-movies/src/main/java/io/fimltime/data/tmdb/movies/TmdbMovieRepositoryImpl.kt (hunk 6 above)
Comment (hadi-norouzi) [reply data/tmdb-movies/src/main/java/io/fimltime/data/tmdb/movies/TmdbMovieRepositoryImpl.kt:66]: I want to store some permanent data of movies like title or description to show very quickly in UI and when Network result arrived I update my database and emit new data.
So if we want to implement it as offline first do you suggest store all details? and in ```GetMovieDetailsUseCase``` observe on database and only request network call in ```TmdbMovieRepository```?

File: gradle/libs.versions.toml (hunk 7 above)
Comment (moallemi) [reply gradle/libs.versions.toml:79]: No Problem. We will migrate to ksp in a separate issue/pr #59

File: data/tmdb-movies/src/main/java/io/fimltime/data/tmdb/movies/TmdbMovieRepositoryImpl.kt (hunk 6 above)
Comment (moallemi) [reply data/tmdb-movies/src/main/java/io/fimltime/data/tmdb/movies/TmdbMovieRepositoryImpl.kt:66]: If you want to use cache for this purpose we have to think more about the implementation.

**scenario:** 
You have a cached data for a movie. User navigates to a movie detail screen again in since we have the cache, we quickly show him the initial title, image and description. But after a few seconds the network call fails. What's your action now? 
**Option A:** 
It's not a good idea (UX) to show an error screen because a few moments ago you were displaying the initial data and suddenly you switch to error screen. 
**Option B**
Show a snack bar or toast indicating that network failed. So what is next for user? is the info on screen enough for him to use or should he refresh the screen? Do you want to add a load again or refresh button in every placed with missing data? (genres or langs, ...)

That's why I suggest to store all data to make the implementation as simple as possible.

You can always observe the database and only propagate the network error result if there is no cached data.

**Comment by moallemi** [comment]
@hadi-norouzi let me know when your PR is ready for review

File: data/tmdb-movies/src/main/java/io/fimltime/data/tmdb/movies/TmdbMovieRepositoryImpl.kt [hunk 8]
```diff
@@ -6,54 +6,77 @@ import androidx.paging.PagingData
...
-          is Result.Failure -> traktIdResult
-          is Result.Success -> {
-            val traktId = traktIdResult.data.toInt()
-            return when (val watched = traktSyncRemoteSource.getHistoryById(traktId.toString())) {
-              is Result.Failure -> result.run {
-                copy(
-                  data = data.copy(
-                    ids = data.ids.copy(
-                      traktId = traktId,
+  override suspend fun getMovieDetails(movieId: Int): Flow<Result<VideoDetail, GeneralError>> {
```
Comment (hadi-norouzi) [inline data/tmdb-movies/src/main/java/io/fimltime/data/tmdb/movies/TmdbMovieRepositoryImpl.kt:31]: @moallemi Could you please review this function and help me how to improve this function? 
It works right now but I want to only depend on database but first time we get movie it returns ```null``` and we show user an error, not loading.
And another thing is that ```GlobalScope.lunch```. I need to lunch a coroutine but not suspend my current function what is your suggestion?

Comment (moallemi) [reply data/tmdb-movies/src/main/java/io/fimltime/data/tmdb/movies/TmdbMovieRepositoryImpl.kt:31]: I think we do not need a GlobalScope for now. one implementation could be:

```kotlin
override suspend fun getMovieDetails(movieId: Int): Flow<Result<VideoDetail, GeneralError>> = flow {
    val movie = movieDao.getMovieByTmdbId(movieId).firstOrNull()
    if (movie != null) {
      emit(Result.Success(movie.toMovie()))
      // TODO: Check if movie is outdated and then fetch from network
      fetchMovieDetailsFromNetwork(movieId)
    } else {
      emit(fetchMovieDetailsFromNetwork(movieId))
    }
  }

  private suspend fun fetchMovieDetailsFromNetwork(movieId: Int): Result<VideoDetail, GeneralError> =
    when (val result = tmdbMoviesRemoteSource.movieDetails(movieId)) {
      is Result.Failure -> result
      is Result.Success -> {
        movieDao.storeMovie(result.data.toEntity())
        when (val traktIdResult = traktMovieSearchRemoteSource.getByTmdbId(result.data.ids.tmdbId.toString())) {
          is Result.Failure -> traktIdResult
          is Result.Success -> {
            val traktId = traktIdResult.data.toInt()
            when (val watched = traktSyncRemoteSource.getHistoryById(traktId.toString())) {
              is Result.Failure -> result.run {
                copy(
                  data = data.copy(
                    ids = data.ids.copy(
                      traktId = traktId,
                    ),
                  ),
                )
              }

              is Result.Success -> {
                result.run {
                  copy(
                    data = data.copy(
                      ids = data.ids.copy(
                        traktId = traktId,
                      ),
                      isWatched = watched.data,
                    ),
                  )
                }
              }
            }
          }
        }
      }
    }
```
We currently don't need to verify if the movie is outdated. We can postpone implementing this feature and incorporate a cache expiration of either one or five days at a later stage.

File: data/database/src/main/java/io/filmtime/data/database/StringListConverter.kt [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+  @TypeConverter
+  fun stringToList(stringList: String): List<String> {
+    return stringList.split(",")
+  }
+
+  @TypeConverter
+  fun listToString(list: List<String>): String {
+    return list.joinToString(",")
+  }
+}
```
Comment (moallemi) [inline data/database/src/main/java/io/filmtime/data/database/StringListConverter.kt:16]: It may seem like I'm splitting hairs, but this approach has some corner cases. e.g: You try to store 
```kotlin
listOf("A,B", "C", "D")
```

and you restore as:
```kotlin
listOf("A", "B", "C", "D")
```

**No need to change it right now.**

File: feature/movie-detail/src/main/java/io/filmtime/feature/movie/detail/MovieDetailScreen.kt [hunk 10]
```diff
@@ -293,11 +293,7 @@ fun MovieDetailContent(
     }
     Text(
       modifier = Modifier.padding(horizontal = 16.dp),
-      style = TextStyle(
-        fontWeight = FontWeight.Bold,
-        fontSize = 16.sp,
-        color = Color.Black,
-      ),
+      style = MaterialTheme.typography.bodyLarge.copy(fontWeight = FontWeight.Bold),
```
Comment (moallemi) [inline feature/movie-detail/src/main/java/io/filmtime/feature/movie/detail/MovieDetailScreen.kt:296]: **Pro Tip:**  💅💅💅 Please do not change something that is not related to this Issue/PR. Maybe someone else is working on this screen at the same time and it might led to conflict.

File: feature/movie-detail/src/main/java/io/filmtime/feature/movie/detail/MovieDetailViewModel.kt [hunk 11]
```diff
@@ -132,44 +133,47 @@ class MovieDetailViewModel @Inject constructor(
   }
 
   fun load() = viewModelScope.launch {
-    _state.value = _state.value.copy(isLoading = true)
-
-    when (val result = getMovieDetail(videoId)) {
-      is Success -> {
-        _state.update { state ->
-          state.copy(videoDetail = result.data, isLoading = false)
+    _state.update { it.copy(isLoading = true) }
```
Comment (moallemi) [inline feature/movie-detail/src/main/java/io/filmtime/feature/movie/detail/MovieDetailViewModel.kt:136]: there is a known bug here that we should reset error as well. otherwise our offline implementation does not work well.

```suggestion
    _state.update { it.copy(isLoading = true, error = null) }
```

File: feature/movie-detail/src/main/java/io/filmtime/feature/movie/detail/MovieDetailViewModel.kt [hunk 12]
```diff
@@ -132,44 +133,47 @@ class MovieDetailViewModel @Inject constructor(
...
+
+    getMovieDetail(videoId).collect {
+      Log.d("repeat", videoId.toString())
```
Comment (moallemi) [inline feature/movie-detail/src/main/java/io/filmtime/feature/movie/detail/MovieDetailViewModel.kt:139]: Please Do not forget to remove this log

**Comment by hadi-norouzi** [comment]
> @hadi-norouzi let me know when your PR is ready for review

It's completed.

**Review by moallemi — APPROVED** [review]
